verification/spec/connection-handshake/updateclient-deadlock/ConnectionHandshakeModule.tla

```diff
@@ -0,0 +1,383 @@
+
+*)
+ValidLocalEnd(para) ==
+    /\ para.localEnd.connectionID \in ConnectionIDs
+    /\ para.localEnd.clientID \in ClientIDs
+
+
+(* Operator for reversing the connection ends.
+
+    Given a ConnectionParameters record 'para', returns a new set
+    of parameters where the local and remote ends are
+    flipped (i.e., reversed).
+ *)
+FlipConnectionParameters(para) ==
+    [localEnd   |-> para.remoteEnd,
+     remoteEnd  |-> para.localEnd]
+
+
+(* Operator for constructing a connection proof.
+
+    The connection proof is used to demonstrate to another chain that the
+    local store on this chain comprises a connection in a certain state. 
+ *)
+GetConnProof(myConnection) ==
+    [connection |-> myConnection]
+
+
+(* Operator for constructing a client proof.
+ *)
+GetClientProof ==
+    [latestHeight |-> store.client.latestHeight,
+     consensusHeights |-> store.client.consensusHeights]
+
+
+(* Verification of a connection proof. 
+
+    This is a state predicate returning true if the following holds: 
+    - the state of connection in this proof should match with input parameter
+    'expectedState'; and
+    - the connection parameters in this proof should match with the flipped version
+    of the input 'expectedParams'.
+
+ *)
+VerifyConnProof(cp, expectedState, expectedParams) ==
+    /\ cp.connection.state = expectedState
+    /\ cp.connection.parameters = FlipConnectionParameters(expectedParams)
+
+
+(* Verification of a client proof.
+
+    This is a state predicate returning true if the following holds: the height
+    reported in the client proof must not exceed the current (latestHeight) of
+    this chain.
+ *)
+VerifyClientProof(cp) ==
+    /\ cp.latestHeight <= store.latestHeight   (* Consistency height check. *)
+    /\ cp.latestHeight \in cp.consensusHeights (* Client verification step. *)
+
+
+(***************************************************************************
+ Connection Handshake Module actions & operators.
+ ***************************************************************************)
+
+
+(* Modifies the local store.
+
+    Replaces the connection in the store with the argument 'newCon'.
+    This action also advances the chain height. 
+ *)
+NewStore(newCon) ==
+    [store EXCEPT !.connection = newCon,
+                  !.latestHeight = @ + 1]
+
+
+(* Handles a "ICS3MsgInit" message 'm'.
+
+    Primes the store.connection to become initialized with the parameters
+    specified in 'm'. Also creates a reply message, enqueued on the outgoing
+    buffer.
+ *)
+HandleInitMsg(m) ==
+    LET newCon == [parameters |-> m.parameters,
+                   state      |-> "INIT"]
+        myConnProof == GetConnProof(newCon)
+        myClientProof == GetClientProof
+        replyMsg == [parameters |-> FlipConnectionParameters(m.parameters),
+                     type |-> "ICS3MsgTry",
+                     proofHeight |-> store.latestHeight,
```

Comment:
Making the comment here but also valid for all other handles. I believe there is a problem here as the proof height is current height but then the store is updated below in `NewStore(newCon)` with +1 height. In real scenario, the store is updated to a height H (where the IBC msg event occurred) and then the relayer takes the proofs at height H. So we want to simulate this here.
If I make this change (and in all other proofs): 
```suggestion
                     proofHeight |-> store.latestHeight + 1,
```
the model stutters. Needs change to maxHeight 6 and then we run into a deadlock scenario somewhat similar to one described here. The trace shows even more clearly that the environment simulates in fact multiple relayers.

Reply:
This is a great find, I was not aware that creating the proof at `store.latestHeight` is in a sense unrealistic.